geode/src/main/scala/akka/stream/alpakka/geode/internal/stage/GeodeCQueryGraphLogic.scala

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.geode.internal.stage
+
+import akka.stream.SourceShape
+import akka.stream.stage.StageLogging
+import org.apache.geode.cache.client.ClientCache
+import org.apache.geode.cache.query.{CqAttributesFactory, CqEvent, CqQuery, Struct}
+import org.apache.geode.cache.util.CqListenerAdapter
+
+abstract class GeodeCQueryGraphLogic[V](val shape: SourceShape[V],
+                                        val clientCache: ClientCache,
+                                        val queryName: Symbol,
+                                        val sql: String)
+    extends GeodeSourceStageLogic[V](shape, clientCache)
+    with StageLogging {
+
+  var query: Option[CqQuery] = None
+
+  override def preStart(): Unit = {
+
+    val cqf = new CqAttributesFactory()
+
+    val eventListener = new CqListenerAdapter() {
+      override def onEvent(ev: CqEvent): Unit =
+        onElement.invoke(ev.getNewValue().asInstanceOf[V])
```

Comment:
Here `onElement` is:

```
if (isAvailable(out)) pushElement(out, element)
else enqueue(element)
```
which is dangerous as it will grow the queue without any bounds.

Here we have an opportunity to block on the Geode library thread and in such way signal backpressure to the Geode library. Take a look at this [blog post](http://blog.akka.io/integrations/2016/09/05/flow-control-at-the-akka-stream-boundary#blocking-the-library-thread) for a more thorough description how this should be done.

Also you can take a look at Mqtt connector here in Alpakka project which uses [exactly the same strategy](https://github.com/akka/alpakka/blob/master/mqtt/src/main/scala/akka/stream/alpakka/mqtt/MqttSourceStage.scala#L63-L66) to signal backpressure to the Mqtt library.